Speed up KeyAnalyser

IDisposable opened the pull request:
While reviewing [#87510 ](https://github.com/dotnet/runtime/pull/87510), I noticed the inline code in the Comparers seems like since it's in the hot-loop path, might be faster to move the `IsLeft` conditionalized code out of the loop by adding a Slicer to the comparator. The Slicer is the same logic as was in the bodies of the `Equals` and `GetHashCode` methods, and indeed also matches the delegate that was passed to the internal `CreateAnalysisResults` method.

NOT SURE if this actually will speed things up at all, hoping to leverage some _/azp_ magic to do the performance evaluations.

IDisposable commented:
/azp help

IDisposable commented:
/azp list

IDisposable commented on src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/KeyAnalyzer.cs:
```diff
@@ -124,7 +120,7 @@ private static bool TryUseSubstring(ReadOnlySpan<string> uniqueStrings, bool ign
                 // was requested, as there's nothing that would compare equally to the substring
                 // other than the substring itself.
-                bool canSwitchIgnoreCaseToCaseSensitive = ignoreCase;
+                bool canSwitchIgnoreCaseToCaseSensitive = true;
```
This is demonstrably `true` because we're inside the [`if (ignoreCase)` immediately above](https://github.com/dotnet/runtime/pull/87590/files#diff-8dfef3756fbf6ca0aab7668e6abf54ae6c416c60cb1b3bd86455204942f7dbefR117)

IDisposable commented:
/azp run runtime-coreclr superpmi-diffs

IDisposable commented:
Hey @adamsitnik, can you do the `/azp` magic to make this run the perf tests?

adamsitnik commented:
> Hey @adamsitnik, can you do the /azp magic to make this run the perf tests?

I am afraid we don't have that in dotnet/runtime yet (cc @sblom)

For now you need to follow the docs I provided in https://github.com/dotnet/runtime/pull/87510#issuecomment-1592438426 and benchmark it locally

IDisposable commented:
Sadly @adamsitnik , I can't even get the dotnet/performance stuff to execute _at all_ on my machine... :( https://github.com/dotnet/performance/issues/3096

IDisposable commented:
Initial results
BenchmarkDotNet=v0.13.2.2052-nightly, OS=Windows 11 (10.0.22631.1906)
Intel Core i7-10875H CPU 2.30GHz, 1 CPU, 16 logical and 8 physical cores
.NET SDK=8.0.100-preview.7.23322.33
  [Host]     : .NET 8.0.0 (8.0.23.32106), X64 RyuJIT AVX2
  Job-RBOOLL : .NET 8.0.0 (42.42.42.42424), X64 RyuJIT AVX2
  Job-GHPBIR : .NET 8.0.0 (42.42.42.42424), X64 RyuJIT AVX2

PowerPlanMode=00000000-0000-0000-0000-000000000000  Arguments=/p:EnableUnsafeBinaryFormatterSerialization=true  IterationTime=250.0000 ms
MaxIterationCount=20  MinIterationCount=15  WarmupCount=1

|           Method |                                                                                                            Toolchain | Size |     Mean |     Error |    StdDev |   Median |      Min |      Max | Ratio | MannWhitney(2ms) | Allocated | Alloc Ratio |
|----------------- | ------------------------------------------------------------------------------------------------------------------- |----- |---------:|----------:|----------:|---------:|---------:|---------:|------:|----------------- |----------:|------------:|
| FrozenDictionary |  Change |  512 | 6.363 us | 0.0315 us | 0.0263 us | 6.364 us | 6.317 us | 6.404 us |  1.00 |  Base |         - |          NA |
| FrozenDictionary | Baseline |  512 | 6.447 us | 0.0726 us | 0.0644 us | 6.438 us | 6.360 us | 6.577 us |  1.01 |   Same |         - |          NA |

IDisposable commented:
So, is 1.01 (old code) vs. 1.00 (new code) worthwhile? It isn't slower, and it (to me) is less coupled.

IDisposable commented on src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/KeyAnalyzer.cs:
```diff
@@ -263,25 +259,32 @@ public AnalysisResults(bool ignoreCase, bool allAsciiIfIgnoreCase, int hashIndex
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ReadOnlySpan<char> SliceLeft(string s, int index, int count) => s!.AsSpan(index, count);
```
Do the slice like we would in https://github.com/dotnet/runtime/pull/87590/files#diff-8dfef3756fbf6ca0aab7668e6abf54ae6c416c60cb1b3bd86455204942f7dbefL100

IDisposable commented on src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/KeyAnalyzer.cs:
```diff
@@ -263,25 +259,32 @@ public AnalysisResults(bool ignoreCase, bool allAsciiIfIgnoreCase, int hashIndex
+        public static ReadOnlySpan<char> SliceLeft(string s, int index, int count) => s!.AsSpan(index, count);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ReadOnlySpan<char> SliceRight(string s, int index, int count) => s!.AsSpan(s!.Length + index, count);
```
This encapsulates the login on [old line 100](https://github.com/dotnet/runtime/pull/87590/files#diff-8dfef3756fbf6ca0aab7668e6abf54ae6c416c60cb1b3bd86455204942f7dbefL100)

IDisposable commented on src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/KeyAnalyzer.cs:
```diff
@@ -263,25 +259,32 @@ public AnalysisResults(bool ignoreCase, bool allAsciiIfIgnoreCase, int hashIndex
         {
-            public override bool Equals(string? x, string? y) => x.AsSpan(IsLeft ? Index : (x!.Length + Index), Count).SequenceEqual(y.AsSpan(IsLeft ? Index : (y!.Length + Index), Count));
-            public override int GetHashCode(string s) => Hashing.GetHashCodeOrdinal(s.AsSpan(IsLeft ? Index : (s.Length + Index), Count));
+            public override bool Equals(string? x, string? y) => Slicer(x!, Index, Count).SequenceEqual(Slicer(y!, Index, Count));
```
Here we can just call Slicer, which and no longer has an IsLeft test in a hot loop, so calling through to the `GetSpan`-like method that was set on lines 60 (for the `IsLeft == true` logic) and 84 (for the `IsLeft == false` logic). This (to my reasoning) will result in much better low-level branch-prediction and since the actual delegate bodies are ` [MethodImpl(MethodImplOptions.AggressiveInlining)]` they should be inlined stubs.

IDisposable replied:
I am *not* confident about the various not-null hints `!` that I injected, but I think they're right, if not overkill. Not sure it really would affect the code generated and welcome input on that from low-level ninjas

IDisposable commented:
I'm not sure this is an endpoint or a jumping off for more or simply an exploration that should be tabled. I welcome input from @adamsitnik and @MichalStrehovsky or anyone else :)

IDisposable committed:
Move the IsLeft/IsRight decision out of the loop

The slicing is really only changed once per Count, so move the
IsLeft-dependent logic into delegates to see if that makes
things faster.

A little more refactoring
Encourage inlining.

IDisposable commented:
Closing this because it's gotten messy
